Refactor kReadCommissioningInfo step in DeviceCommissioner

ksperling-apple opened the pull request:
This PR refactors the kReadCommissioningInfo DeviceCommissioner step in preparation for adding the ability to read additional data. Having a fixed split of the attributes to be read into two steps in inflexible and makes it hard to optimize the reads that are performed.

Specifically this change
- Rolls kReadCommissioningInfo2 stage into kReadCommissioningInfo and tracks progress internally instead
- Groups attributes being read and parsing methods by cluster (where this wasn't already done)
- Makes the parsing code more consistent and simplifies some logic.

If desired this PR can be reviewed commit by commit; builds and tests were green after each commit.

ksperling-apple committed:
Roll kReadCommissioningInfo2 stage into kReadCommissioningInfo and track progress internally

bzbarsky-apple commented on src/controller/CHIPDeviceController.cpp:
```diff
@@ -2189,29 +2190,110 @@ void DeviceCommissioner::OnDeviceConnectionRetryFn(void * context, const ScopedN
+    app::AttributePathParams readPaths[InteractionModelEngine::kMinSupportedPathsPerReadRequest];
+    static_assert(InteractionModelEngine::kMinSupportedPathsPerReadRequest >= 9);
+
+    const auto proxy    = mDeviceBeingCommissioned;
```
So what happens if we were in the middle of a read and commissioning got canceled?  I think proxy here will be null, right?  

In general, the handling of things that happen during the async parts of this process needs some careful thinking about.  It was handled by the general "go to next step" machinery before, but now we have to make sure it's correct.

ksperling-apple replied:
Hm, that was true for the initial `kReadCommissioningInfo` step, but I'm not sure cancellation during `kRead...2` was actually handled correctly. In `CancelCommissioningInteractions` we kill the ReadClient, and this calls `OnError` and `OnDone`. Only `OnDone` is currently implemented, and `ParseCommissioningInfo` will then probably happen to fail... Maybe `CancelCommissioningInteractions` needs to just finish the step first, and then we can just ignore the `OnDone` call.

ksperling-apple replied:
Actually that's not quite correct... the ReadClient destructor won't call `OnDone`, so I don't think there is anything different to be done here about how we handle cancellation before and after my change.

ksperling-apple replied:
Just to elaborate a bit more, cancellation calls `CancelCommissioningInteractions()`, we free the ReadClient (and now also the ClusterStateCache, this was missing before), and `OnDone` won't be called. So we won't get back into `ContinueReadingCommissioningInfo` or touch `mDeviceBeingCommissioned` in that case.

bzbarsky-apple commented on src/controller/CHIPDeviceController.cpp:
```diff
@@ -2916,85 +2998,23 @@ void DeviceCommissioner::PerformCommissioningStep(DeviceProxy * proxy, Commissio
+        // Allocate a ClusterStateCache to collect the data from our read requests.
         // The cache will be released in:
         // - SendCommissioningReadRequest when failing to send a read request.
         // - ParseCommissioningInfo when the last ReadCommissioningInfo stage is completed.
```
There is no ParseCommissioningInfo anymore.

ksperling-apple committed:
Fix trigger-based cancellation in MTRPairingTests

As of #32974 we were potentially cancelling later than intended.
Also avoid re-using KVS file names to avoid pollution by earlier failed runs.

ksperling-apple committed:
Refactor ContinueReadingCommissioningInfo to group attributes dynamically

Also group requests by cluster. The logic of which attributes to read is unchanged.

ksperling-apple committed:
Refactor FinishReadingCommissioningInfo / Parse*Info

Group by cluster instead of by the stages that no longer exist.
Also align error handling and logging a little.

ksperling-apple committed:
Simplify ParseNetworkCommissioningInfo

- Use ClusterStateCache::Get variant that decodes directly
- No need to loop again to parse ConnectMaxTimeSeconds attributes
- Per spec Ethernet does not have ConnectMaxTimeSeconds

tcarmelveilleux commented on src/controller/CHIPDeviceController.cpp:
```diff
@@ -2189,46 +2194,166 @@ void DeviceCommissioner::OnDeviceConnectionRetryFn(void * context, const ScopedN
+    bool AddAttributePath(Ts &&... args)
     {
-        err = ParseCommissioningInfo2(info);
+        VerifyOrReturnValue(mSkip == 0, true, mSkip--);
+        VerifyOrReturnValue(mCount < kCapacity, false, mCount = kCapacity + 1);
```
These should be if's . The 3 complex args are really really hard to parse on a single read.

ksperling-apple replied:
Done

tcarmelveilleux commented on src/controller/CHIPDeviceController.cpp:
```diff
@@ -2189,46 +2194,166 @@ void DeviceCommissioner::OnDeviceConnectionRetryFn(void * context, const ScopedN
+    ReadInteractionBuilder builder(mReadCommissioningInfoProgress);
+    [&]() -> void {
+        // General Commissioning
+        VerifyOrReturn(builder.AddAttributePath(kRootEndpointId, Clusters::GeneralCommissioning::Id,
+                                                Clusters::GeneralCommissioning::Attributes::SupportsConcurrentConnection::Id));
+        VerifyOrReturn(builder.AddAttributePath(kRootEndpointId, Clusters::GeneralCommissioning::Id,
+                                                Clusters::GeneralCommissioning::Attributes::Breadcrumb::Id));
+        VerifyOrReturn(builder.AddAttributePath(kRootEndpointId, Clusters::GeneralCommissioning::Id,
+                                                Clusters::GeneralCommissioning::Attributes::BasicCommissioningInfo::Id));
+        VerifyOrReturn(builder.AddAttributePath(kRootEndpointId, Clusters::GeneralCommissioning::Id,
+                                                Clusters::GeneralCommissioning::Attributes::RegulatoryConfig::Id));
+        VerifyOrReturn(builder.AddAttributePath(kRootEndpointId, Clusters::GeneralCommissioning::Id,
+                                                Clusters::GeneralCommissioning::Attributes::LocationCapability::Id));
```
A wildcard reead of general commissioning is probably small enough to get all you need and reduce number of paths there.

ksperling-apple replied:
Yes, that's one of the optimizations I'm looking up for a follow-up. For ICD we can also do better I think, either by reading just the feature flag first and the rest during the second read (if any), or as a wildcard. But I wanted to keep those behavior changes separate from the refactoring.

tcarmelveilleux commented on src/controller/CHIPDeviceController.cpp:
```diff
@@ -2189,46 +2194,166 @@ void DeviceCommissioner::OnDeviceConnectionRetryFn(void * context, const ScopedN
+                                                    Clusters::OperationalCredentials::Attributes::Fabrics::Id));
+        }
+
+        // ICD Management
+        if (params.GetICDRegistrationStrategy() != ICDRegistrationStrategy::kIgnore)
+        {
+            VerifyOrReturn(builder.AddAttributePath(kRootEndpointId, Clusters::IcdManagement::Id,
+                                                    Clusters::IcdManagement::Attributes::FeatureMap::Id));
+        }
```
Here a wildcard would also help.

tcarmelveilleux commented on src/controller/CHIPDeviceController.cpp:
```diff
@@ -2189,46 +2194,166 @@ void DeviceCommissioner::OnDeviceConnectionRetryFn(void * context, const ScopedN
+}
+
+namespace {
+void AccumulateErrors(CHIP_ERROR & acc, CHIP_ERROR err)
```
Suggest `AccumulateEarlierError`

ksperling-apple replied:
Hm, I'm not sure the "earlier" really adds clarity, I think I'll leave it as is for now.

tcarmelveilleux commented on src/controller/CHIPDeviceController.cpp:
```diff
@@ -2238,247 +2363,228 @@ void DeviceCommissioner::ParseCommissioningInfo()
             {
-                info.network.eth.minConnectionTime = time;
+                ChipLogProgress(Controller, "NetworkCommissioning Features: no features.");
+                // TODO: Gross workaround for the empty feature map on all clusters. Remove.
```
That is a huge bug in all clusters. The right thing to do is to fix it.

ksperling-apple replied:
Yeah, I'll have a look if this is still needed (the workaround is existing code).

cecille replied:
I'm not sure this even happens any more - do you know if you're actually getting a blank feature map back?

ksperling-apple replied:
Yeah the current version of network-commissioning-server never returns a 0 feature map, I'll remove this.

tcarmelveilleux commented on src/controller/python/test/test_scripts/commissioning_failure_test.py:
```diff
@@ -106,7 +107,8 @@ async def main():
                       "Commissioning failure tests failed for simulated report failure on stage {}".format(testFailureStage))
 
     else:
-        for testFailureStage in range(3, 21):
+        # [kArmFailsafe, kConfigureTrustedTimeSource] (inclusive). TODO: https://github.com/project-chip/connectedhomeip/issues/36629
+        for testFailureStage in range(3, 20):
```
This range change is odd. I also do not understand  the command

ksperling-apple replied:
The range being iterated is kArmFailsafe to kConfigureTrustedTimeSource (inclusive), before and after. The numbers changed because I removed kReadCommissioningInfo2. I added the todo because we really should be exposing these as named constants to Python somehow.

ksperling-apple commented on src/python_testing/TC_CGEN_2_4.py:
```diff
@@ -50,14 +50,15 @@
-kSendOpCertSigningRequest = 15
-kSendTrustedRootCert = 18
-kSendNOC = 19
+# TODO: https://github.com/project-chip/connectedhomeip/issues/36629
+kArmFailsafe = 3
+kConfigRegulatory = 4
+kSendPAICertificateRequest = 9
+kSendDACCertificateRequest = 10
+kSendAttestationRequest = 11
+kSendOpCertSigningRequest = 14
+kSendTrustedRootCert = 17
+kSendNOC = 18
```
Indeed, that's why I filed that issue.

ksperling-apple committed:
Review comments: Use explicit ifs for this logic

cecille reviewed (commented):
This generally seems better to me. Most of my comments are minor. +1 to Boris' question about cancellation though.

ksperling-apple commented on src/controller/CHIPDeviceController.cpp:
```diff
@@ -2189,46 +2194,175 @@ void DeviceCommissioner::OnDeviceConnectionRetryFn(void * context, const ScopedN
+    bool AddAttributePath(Ts &&... args)
     {
-        err = ParseCommissioningInfo2(info);
+        if (mSkip > 0)
```
Yes that’s correct, will add a note

ksperling-apple commented on src/controller/CHIPDeviceController.cpp:
```diff
@@ -2189,46 +2194,175 @@ void DeviceCommissioner::OnDeviceConnectionRetryFn(void * context, const ScopedN
+        // Keep track of the number of attributes we have read already so we can resume from there.
+        auto progress = mReadCommissioningInfoProgress + builder.size();
+        VerifyOrDie(progress < kReadProgressNoFurtherAttributes);
+        mReadCommissioningInfoProgress = static_cast<decltype(mReadCommissioningInfoProgress)>(progress);
```
Doing the math with uint8 is annoying due to promotion to int, so having the logic use size_t is convenient. At the same time I wanted to be a bit more frugal with what gets stored in the DeviceCommissioner permanently.

ksperling-apple committed:
Review comments: Explain skip logic

ksperling-apple committed:
Remove workaround for zero network commissioning feature map

bzbarsky-apple commented on src/darwin/Framework/CHIPTests/TestHelpers/MTRTestCase+ServerAppRunner.m:
```diff
@@ -61,7 +61,7 @@ + (NSTask *)doStartAppWithName:(NSString *)name arguments:(NSArray<NSString *> *
         [NSString stringWithFormat:@"%llu", passcode.unsignedLongLongValue],
         @"--KVS",
-        [NSString stringWithFormat:@"/tmp/chip-%@-kvs%u", name, uniqueIndex],
+        [NSString stringWithFormat:@"/tmp/xctest-%d-chip-%@-kvs%u", getpid(), name, uniqueIndex],
```
We should consider making a similar change for the outFile/errorFile bits below.  Followup is fine for that.

ksperling-apple replied:
We could. For logs it's not really a problem to overwrite the existing file, but for the KVS we don't overwrite.

bzbarsky-apple commented on src/controller/CHIPDeviceController.h:
```diff
@@ -839,7 +839,9 @@ class DLL_EXPORT DeviceCommissioner : public DeviceController,
 
     CommissioningStage mCommissioningStage = CommissioningStage::kSecurePairing;
-    bool mRunCommissioningAfterConnection  = false;
+    uint8_t mReadCommissioningInfoProgress = 0;
```
Document that this tracks the number of attribute paths that had reads issued so far?

ksperling-apple committed:
Review comment: document mReadCommissioningInfoProgress

bzbarsky-apple commented on src/controller/CHIPDeviceController.cpp:
```diff
@@ -2259,6 +2259,7 @@ void DeviceCommissioner::ContinueReadingCommissioningInfo(const CommissioningPar
 {
     VerifyOrDie(mCommissioningStage == CommissioningStage::kReadCommissioningInfo);
 
+    // mReadCommissioningInfoProgress starts at 0 and counts the number of attributes we have read.
```
This is specifically the number of attribute paths we have issued reads for, not the number of attribute values we got back, due to wildcarding...

ksperling-apple committed:
Review comment: avoid paths / attributes ambiguity
